File: services/graphql-crm/src/models/invoice_items.rs
use std::sync::Arc;

use async_graphql::{ComplexObject, Context, SimpleObject, dataloader::Loader};
use chrono::{DateTime, Utc};
use fake::Dummy;
use graphql_core::PostgresDataLoader;
use rust_decimal::Decimal;
use sqlx::FromRow;
use uuid::Uuid;

use crate::models::{invoices, products};

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, Eq, SimpleObject, FromRow, Dummy)]
#[graphql(name = "CrmInvoiceItems", complex)]
pub struct Model {
    pub id: Uuid,
    #[graphql(skip)]
    pub invoice_id: Uuid,
    #[graphql(skip)]
    pub product_id: Uuid,
    pub quantity: i32,
    pub price: Decimal,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[ComplexObject]
impl Model {
    async fn invoice(&self, ctx: &Context<'_>) -> async_graphql::Result<invoices::Model> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        Ok(loader
            .load_one(invoices::PrimaryKey(self.invoice_id))
            .await?
            .ok_or(async_graphql::Error::new("Unable to find product"))?)
    }
    async fn product(&self, ctx: &Context<'_>) -> async_graphql::Result<products::Model> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        Ok(loader
            .load_one(products::PrimaryKey(self.product_id))
            .await?
            .ok_or(async_graphql::Error::new("Unable to find product"))?)
    }
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        let keys = keys.iter().map(|k| k.0).collect::<Vec<_>>();

        let results =
            sqlx::query_as::<_, Self::Value>("select * from crm.invoice_items where id = ANY($1)")
                .bind(&keys)
                .fetch_all(&self.pool)
                .await?
                .into_iter()
                .map(|model| (PrimaryKey(model.id), model))
                .collect::<_>();

        Ok(results)
    }
}

File: packages/graphql/src/schema/crm/invoice_items/resolvers/InvoiceItems.ts
import type { Invoices, Products } from "../../../../zod.schema";
import type { InvoiceItemsResolvers } from "./../../../types.generated";
export const InvoiceItems: InvoiceItemsResolvers = {
	invoice: async (parent, _args, ctx) => {
		const result = await ctx.db
			.selectFrom("crm.invoices")
			.selectAll("crm.invoices")
			.innerJoin(
				"crm.invoiceItems",
				"crm.invoiceItems.invoiceId",
				"crm.invoices.id",
			)
			.where("crm.invoiceItems.id", "=", parent.id as string)
			.executeTakeFirst();

		return result as unknown as Invoices;
	},
	product: async (parent, _args, ctx) => {
		const result = await ctx.db
			.selectFrom("crm.products")
			.selectAll("crm.products")
			.innerJoin(
				"crm.invoiceItems",
				"crm.invoiceItems.productId",
				"crm.products.id",
			)
			.where("crm.invoiceItems.id", "=", parent.id as string)
			.executeTakeFirst();

		return result as unknown as Products;
	},
};
